### PR TITLE
Admin API decorators 

### PR DESCRIPTION
**Pull request by sravfeyn**

### CODE REVIEW

**Comment by sravfeyn** [comment]
Hey @esoergel I have been trying to debug this build failure without much luck. Do you have any insights?
```
======================================================================
ERROR: test_basic (corehq.apps.api.tests.AdminAPITest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/commcare-hq-ro/corehq/apps/api/tests.py", line 1699, in test_basic
    self.assert_admin_resource(url)
  File "/mnt/commcare-hq-ro/corehq/apps/api/tests.py", line 1683, in assert_admin_resource
    response = self._assert_auth_get_resource(url, username='admin', password='pass')
  File "/mnt/commcare-hq-ro/corehq/apps/api/tests.py", line 187, in _assert_auth_get_resource
    response = self.client.get(api_url)
  File "/vendor/lib/python2.7/site-packages/django/test/client.py", line 500, in get
    **extra)
  File "/vendor/lib/python2.7/site-packages/django/test/client.py", line 303, in get
    return self.generic('GET', path, secure=secure, **r)
  File "/vendor/lib/python2.7/site-packages/django/test/client.py", line 379, in generic
    return self.request(**r)
  File "/vendor/lib/python2.7/site-packages/django/test/client.py", line 466, in request
    six.reraise(*exc_info)
  File "/vendor/lib/python2.7/site-packages/django/core/handlers/base.py", line 132, in get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/vendor/lib/python2.7/site-packages/django/views/decorators/csrf.py", line 58, in wrapped_view
    return view_func(*args, **kwargs)
  File "/vendor/lib/python2.7/site-packages/tastypie/resources.py", line 219, in wrapper
    response = callback(request, *args, **kwargs)
  File "/vendor/lib/python2.7/site-packages/tastypie/resources.py", line 450, in dispatch_list
    return self.dispatch('list', request, **kwargs)
  File "/mnt/commcare-hq-ro/corehq/apps/api/domain_metadata.py", line 41, in dispatch
    return Resource.dispatch(self, request_type, request, **kwargs)
  File "/vendor/lib/python2.7/site-packages/tastypie/resources.py", line 482, in dispatch
    response = method(request, **kwargs)
  File "/vendor/lib/python2.7/site-packages/tastypie/resources.py", line 1335, in get_list
    to_be_serialized = paginator.page()
  File "/vendor/lib/python2.7/site-packages/tastypie/paginator.py", line 194, in page
    count = self.get_count()
  File "/vendor/lib/python2.7/site-packages/tastypie/paginator.py", line 126, in get_count
    return self.objects.count()
  File "/mnt/commcare-hq-ro/corehq/apps/api/domain_metadata.py", line 26, in count
    return self.es_query.size(0).run().total
  File "/mnt/commcare-hq-ro/corehq/apps/es/es_query.py", line 210, in run
    raw = run_query(self.index, self.raw_query, debug_host=self.debug_host)
  File "/mnt/commcare-hq-ro/corehq/elastic.py", line 166, in run_query
    raise ESError(e)
ESError: TransportError(404, u'IndexMissingException[[test_hqdomains_2016-08-08] missing]')
```

**Comment by czue** [comment]
@sravfeyn looks like the ES index isn't created. you can probably add something like:
```python
initialize_index_and_mapping(get_es_new(), DOMAIN_INDEX_INFO)
```
to setup

**Comment by sravfeyn** [comment]
Thanks cory, that fixed it!

This is a new version of https://github.com/dimagi/commcare-hq/pull/12470, I am sorry I ended up opening a functionally equivalent copy of the old PR instead of making that PR work, because of multiple simultaneous refactors by others, it was difficult to keep that PR in shape.

This PR 
1. Introduces superuser API-exclusive decorators as opposed to decorators that allow session based access
2. Uses these decorators to do superuser API auth
3. Moves auth related classes/utils to its own file

@czue has context.
Buddies @esoergel @kaapstorm

File: corehq/apps/api/tests.py [hunk 1]
```diff
@@ -1759,8 +1752,8 @@ class TestConfigurableReportDataResource(APIResourceTest):
     resource = v0_5.ConfigurableReportDataResource
     api_name = "v0.5"
 
-    @classmethod
-    def _get_list_endpoint(cls):
+    @property
+    def list_endpoint(cls):
```
Comment (kaapstorm) [inline corehq/apps/api/tests.py]: very nitty: `self`

**Comment by kaapstorm** [comment]
(Aside from petty nit) LGTM.

**Comment by kaapstorm** [comment]
:bowtie: You are very tolerant of OCD, @sravfeyn. I'll let @czue merge in case there is anything context-related I might have missed.

File: corehq/apps/api/resources/v0_1.py [hunk 2]
```diff
@@ -4,14 +4,19 @@
 from tastypie import fields
 from tastypie.exceptions import BadRequest
 
+# CCHQ imports
+from corehq.apps.es import FormES
+from corehq.apps.groups.models import Group
+from corehq.apps.users.models import CommCareUser, WebUser, Permissions
```
Comment (czue) [inline corehq/apps/api/resources/v0_1.py]: why are all of these added?

File: corehq/apps/api/resources/v0_4.py [hunk 3]
```diff
@@ -5,10 +5,11 @@
 from tastypie.authentication import Authentication
 from tastypie.exceptions import BadRequest
 
+from corehq.apps.api.auth import CustomResourceMeta, RequirePermissionAuthentication, \
+    DomainAdminAuthentication, api_access_allowed
 from corehq.apps.api.models import ESXFormInstance, ESCase
-from corehq.apps.api.resources.auth import DomainAdminAuthentication, RequirePermissionAuthentication
 from corehq.apps.api.resources.v0_1 import _safe_bool
-from corehq.apps.api.resources.meta import CustomResourceMeta
+from corehq.form_processor.interfaces.dbaccessors import CaseAccessors
```
Comment (czue) [inline corehq/apps/api/resources/v0_4.py]: same - why is this changing?

File: corehq/apps/api/tests.py [hunk 4]
```diff
@@ -1667,6 +1675,46 @@ def test_basic(self):
...
+        community_domain = Domain.get_or_create_with_name('dm', is_active=True)
+        new_user = WebUser.create(community_domain.name, 'admin', 'pass', is_superuser=True)
+        new_user.save()
+
+        self.addCleanup(community_domain.delete)
+        self.addCleanup(new_user.delete)
+
+        for resource in ADMIN_API_LIST:
+            if resource is accounting.DefaultProductPlanResource:
+                # this reource has a bug
```
Comment (czue) [inline corehq/apps/api/tests.py]: nit: this comment would probably be more useful if it explained a bit more about the bug

File: corehq/apps/domain/decorators.py [hunk 5]
```diff
@@ -147,6 +147,36 @@ def wrapper(request, *args, **kwargs):
...
+                        "based on _login_or_challenge for domain-specific auth and validation"
+                    )
+
+            @check_lockout
+            @challenge_fn
+            @require_superuser_or_developer
+            def _authenticate_fn(request, *args, **kwargs):
+                request.couch_user = couch_user = CouchUser.from_django_user(request.user)
+                if couch_user:
+                    clear_login_attempts(couch_user)
```
Comment (czue) [inline corehq/apps/domain/decorators.py]: can you explain why this is needed?

**Review by czue — COMMENTED** [review]
blast from the past.

had a few small questions. also assume you're good with QA?

**Comment by czue** [comment]
@sravfeyn I know you're busy with NIC stuff so just ping me whenever you have a chance to return to this and respond to questions - I don't think any rush to get this in.

**Comment by snopoke** [comment]
Closing old PRs. Reopen if necessary.

**Pull request closed without merging**